chapter09/mssw/exercise1.pl

```diff
@@ -0,0 +1,35 @@
+#!/usr/bin/perl
+use utf8;
+use strict;
+use warnings;
+
+use v5.12;
+
+my %patterns = (
+    Gilligan   => qr/(?:Wiley )?Gilligan/,
+    'Mary-Ann' => qr/Mary-Ann/,
+    Ginger     => qr/Ginger/,
+    Professor  => qr/(?:The )?Professor/,
+    Skipper    => qr/Skipper/,
+    'A Howell' => qr/Mrs?. Howell/,
+);
+
+my $key = rightmost( 'There is Mrs. Howell, Ginger, and Gilligan', \%patterns );
+
+print "Rightmost mathcing key: $key\n";
+
+sub rightmost {
+    my ( $string, $patterns ) = @_;
+
+    my $rightmost = -1;
+    my $rightmost_key = 'Unmatched!';
+    foreach my $key ( keys %$patterns ) {
+        my $position = $string =~ m/$patterns->{$key}/ ? $-[0] : -1;
```

Comment:
The `each` solution the book gives you for the original array of patterns is a good solution for this too - it removes the need to look up the value in the loop body, which would let this line reduce to `my $position = $string =~ $regex ? $-[0] : -1;`.